The garbage collector and quota counter should use the metadata client and protobuf to access resources

smarterclayton opened the pull request:
The garbage collector and the generic object count quota controller currently uses dynamic client and JSON to perform operations only on resource metadata. This PR changes the both to use the metadata client - a client that uses the server-side conversion logic to retrieve only the metadata about the object in protobuf - to more efficiently interact with the server.

Renames GenericInformerFactory to make it clear that these informers return only a subset of objects.

Built on #77819 and #78743 

Part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190322-server-side-get-to-ga.md

/kind cleanup

```release-note
The garbage collector and generic object quota controller have been updated to use the metadata client which improves memory
and CPU usage of the Kube controller manager.
```

smarterclayton committed:
Switch the garbage collector to use metadata client and protobuf

fedebongio commented:
/assign @liggitt @caesarxuchao

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/testing/fixture.go:
```diff
@@ -318,6 +318,11 @@ func (t *tracker) Add(obj runtime.Object) error {
 	}
+
+	if partial, ok := obj.(*metav1.PartialObjectMetadata); ok && len(partial.TypeMeta.APIVersion) > 0 {
+		gvks = []schema.GroupVersionKind{partial.TypeMeta.GroupVersionKind()}
```
I'm confused... does a PartialObjectMetadata have APIVersion/Kind set to `metav1.k8s.io/v1`, `PartialObjectMetadata` or to something like `apps/v1`, `Deployment`?

liggitt replied:
sorry, I'm mixing up PartialObjectMetadata and Table output

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/metadata/fake/simple.go:
```diff
@@ -0,0 +1,390 @@
+// does not expose create
+type MetadataClient interface {
+	metadata.Getter
+	Create(obj *metav1.PartialObjectMetadata, opts metav1.CreateOptions, subresources ...string) (*metav1.PartialObjectMetadata, error)
```
this is a bit weird.... can we name these so they clearly only apply to the fake client (`CreateFake`, `UpdateFake`, etc)

smarterclayton replied:
Done, two reviewers independently suggesting it must be right

liggitt commented on staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go:
```diff
@@ -0,0 +1,156 @@
+)
+
+// NewSharedInformerFactory constructs a new instance of metadataSharedInformerFactory for all namespaces.
+func NewSharedInformerFactory(client metadata.Interface, defaultResync time.Duration) SharedInformerFactory {
```
will this suffer from the same thing as https://github.com/kubernetes/kubernetes/issues/79610?

liggitt replied:
if GC is using partialobjectmeta, and quota is using dynamic, will that actually be worse than both using dynamic?

smarterclayton replied:
Quota would be switched to partial too (for anything count related).

#79610 definitely would apply, this is a straight copy of that code.

yue9944882 replied:
namespace-deletion controller will also switch to partial i suppose?

RainbowMango replied:
> namespace-deletion controller will also switch to partial i suppose?

I think so. refer to [enhancement](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190322-server-side-get-to-ga.md#Motivation)

`and also update controllers that would benefit from use of PartialObjectMetadata (garbage collector, namespace controller, and quota counter) without fear of deprecation. PartialObjectMetadata allows these controllers to perform protobuf list actions efficiently when retrieving only the core object metadata.`

smarterclayton replied:
Namespace deletion is #78744

liggitt commented on pkg/controller/garbagecollector/patch.go:
```diff
@@ -72,6 +72,15 @@ func (gc *GarbageCollector) getMetadata(apiVersion, kind, namespace, name string
 	return meta.Accessor(obj)
 }
 
+type objectForFinalizersPatch struct {
```
clever

smarterclayton replied:
I read this as both positive and negative, as I am sure it was intended.

smarterclayton commented:
Updated, PTAL

liggitt commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -434,7 +435,7 @@ func startGarbageCollectorController(ctx ControllerContext) (http.Handler, bool,
+		metadataClient,
 		ctx.RESTMapper,
 		deletableResources,
 		ignoredResources,
```
this is still passing in `ctx.GenericInformerFactory`, which is still using `dynamicInformers`, not metadata informers

smarterclayton replied:
Renamed and replaced, added a comment about why the name changed.  No object count quota unit tests it looks like, may add those in a follow up (although the logic is super simple since it just observes existing object counts, doesn't even crack open the object).

smarterclayton committed:
Rename metadata.NewConfigOrDie to be consistent

Updated name to match dynamic client

smarterclayton committed:
Use metadata informers instead of dynamic informers in controller manager

All controllers in controller-manager that deal with objects generically
work with those objects without needing the full object. Update the GC
and quota controller to use PartialObjectMetadata input objects which
is faster and more efficient.

smarterclayton commented:
Ok, this is green.  Will run e2e-gce a couple of times looking for flakes, but haven't seen any see far.  Two new commits, rebased on top of master which merged the informer/fake code from the NS PR.

/test pull-kubernetes-e2e-gce

liggitt commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -323,7 +325,7 @@ func (gc *GarbageCollector) attemptToDeleteWorker() bool {
 // returns its latest state.
 func (gc *GarbageCollector) isDangling(reference metav1.OwnerReference, item *node) (
-	dangling bool, owner *unstructured.Unstructured, err error) {
+	dangling bool, owner *metav1beta1.PartialObjectMetadata, err error) {
```
I expected `metav1.PartialObjectMetadata` throughout

liggitt commented:
lgtm, just the one comment about metav1.PartialObjectMetadata vs metav1beta1

liggitt commented:
/lgtm
/hold

feel free to retag once the version is changed

liggitt commented:
/lgtm
/retest
